versions/ufo3/groups.plist.md

```diff
@@ -57,6 +57,10 @@ Now, an application wants to apply the kerning for the pair "AT". There are diff
 
 Two kerning values are found: -50 and -100. There is no way to determine which one is correct. This ambiguity is eliminated by requiring that glyphs occur in only one group per kerning side.
 
+#### 5. Glyphs must not appear more than once in a single kerning group.
+
+Having duplicates of a glyph in a single kerning group adds no extra information and is redundant.
```

Comment:
I'm not so sure about the wording here, it's a bit too informal. Perhaps something along the lines of "Glyphs appearing more than once in the same kerning group causes authoring tool confusion". Not sure if that's best, but something a bit more formal is.

Reply:
In this case it's the production pipeline that gets confused, so "authoring tools" may not be the right thing, either.

That said: are we sure this deserves an error? The other thing that makes kern groups stand out is that their order is irrelevant, and the final order is most likely determined by requirements in the OT spec. Conceptually, kern groups are sets, and any duplicates can safely be ignored, similarly in how duplicates in the glyph order are ignored. In Python, duplicate items are allowed in set literals:
```python
s = {"a", "b", "a"}  # results in {'a', 'b'}
```
I'm not 100% opposed to an error, but it feels a bit heavy for a minor offense. A warning is perhaps nice, but I'm fine with just silently ignoring this situation.

In terms of the UFO spec, I would say that kern groups "should not contain duplicates" rather than "must not contain duplicates".

Reply:
I have 2 comments:

* In case a hard error is too much, would it be acceptable to print a warning when reading such duplicates, and to remove them as early as possible in the UFO reading code, so that client libraries never see the problem? That would be what Python does  if I understand your point.

* I would argue that the order within the group can matter, for example because of the FontLab key glyph situation (at Dalton Maag we use UFOs as a storage format when working with FontLab 5 and it's important that at least the first glyph of the kerning group remains the same)